V2.0 node based

Sploder12 opened the pull request:
Proposed changes
-------------------------------------------
Storing all node connections inside the Game requires unnecessary dictionary lookups and may make code harder to read.
Proposes storing Node connections inside the nodes themselves.

Additionally includes type hints for all current code.
Fixes bug that allows players to enter numbers outside bounds.

Types of changes
-------------------------------------------
- [X] Bugfix (non-breaking change which fixes an issue)
- [x] Quality of Life (Code Readability related changes)

Checklist
-------------------------------------------
- [X] My changeset covers only what is described above (no extraneous changes)
- [X] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)
- [X] Any dependent changes have been merged downstream

Further comments
-------------------------------------------
Type hints will be helpful for linting/auto-suggest/reasoning. I believe we should use them when possible moving forward.

Sploder12 committed:
Added types

Type hints are the greatest thing to grace Python

Sploder12 committed:
Nodes hold connections

Previous required several dictionary lookups, this should be faster and more readable. Additionally some redundant if conditions were removed. Fixed bug regarding invalid number input.

emilkovacev commented on builtin.py:
```diff
@@ -11,13 +12,17 @@ def summary(self):
 
 class Node:
-    def __init__(self, title, desc):
+    def __init__(self, title: str, desc: str, connections: list[Node] = []):
```
Great addition! I agree, type checking in Python is definitely something we should enforce moving forward.

emilkovacev commented on builtin.py:
```diff
@@ -11,13 +12,17 @@ def summary(self):
+    def __init__(self, title: str, desc: str, connections: list[Node] = []):
         self.title = title
         self.desc = desc
+        self.connections = connections
```
I like the idea behind this change, and I see how you tried to keep the API intact while implementing it. My only issue with it is that it interferes with the constructor parameters for the class, which was by design something I attempted to avoid. 

That being said, if we can get it working with another PR without many changes to the file parser, then maybe it's worth it!

Sploder12 replied:
I changed the constructor parameters back to how they were. No changes to the parser should be needed.

emilkovacev replied:
Perfect, that fixes everything. This is ready to merge now.

emilkovacev commented on crpg.py:
```diff
@@ -5,60 +5,61 @@
 
-        # adding node_a to list of connections
-        self.connections[node_a] = self.connections.get(node_a, []) + [node_b]
+        node_a.add_connection(node_b)
```
If we can get the node connections to work, I really like the way you exposed the API to the game

emilkovacev commented on crpg.py:
```diff
@@ -5,60 +5,61 @@
+                continue 
+
+            # check for valid choice
+            if (choice <= len(_next) and choice > 0):
                 break
+            else:
+                print('Please enter a valid number. Try again.')
```
Good catch!

emilkovacev commented on crpg.py:
```diff
@@ -5,60 +5,61 @@
             exit(0)
 
-        _next = self.connections[self.current]
-        if len(_next) > 0: 
```
Good catch on removing this!

emilkovacev reviewed (commented):
All these changes are really good! Adding type hints is incredibly valuable. Once we can get the filetype working, or discuss other options that work better with this implementation, we can merge.

I'm also interested to discuss the main part of this PR, adding the connections into the Nodes.

Sploder12 committed:
Removed Node construction parameter

Having this argument in the constructor doesn't add value and gets in the way of Game handling all connections.
